packages/snaps-execution-environments/src/common/BaseSnapExecutor.ava.test.ts

```diff
@@ -0,0 +1,223 @@
+
+        try {
+          result.hasMethods = {
+            ethereum: {
+              request: 'request' in ethereum,
+              on: 'on' in ethereum,
+              removeListener: 'removeListener' in ethereum,
+              rpcEngine: '_rpcEngine' in ethereum,
+            },
+            snap: {
+              request: 'request' in snap,
+              on: 'on' in snap,
+              removeListener: 'removeListener' in snap,
+              rpcEngine: '_rpcEngine' in snap,
+              requestType: typeof snap.request,
+            }
+          }
+        } catch (error) {
+          result.errors.push(error.message);
+        }
+
+        return result;
+      }
+  `;
+  const executor = new TestSnapExecutor();
+
+  await executor.executeSnap(1, FAKE_SNAP_NAME, CODE, ['ethereum']);
+
+  expect.like(await executor.readCommand(), {
+    jsonrpc: '2.0',
+    id: 1,
+    result: 'OK',
+  });
+
+  await executor.writeCommand({
+    jsonrpc: '2.0',
+    id: 2,
+    method: 'snapRpc',
+    params: [
+      FAKE_SNAP_NAME,
+      ON_RPC_REQUEST,
+      FAKE_ORIGIN,
+      { jsonrpc: '2.0', method: '', params: [] },
+    ],
+  });
+
+  expect.like(await executor.readCommand(), {
+    jsonrpc: '2.0',
+    id: 2,
+    result: {
+      hasMethods: {
+        ethereum: {
+          request: true,
+          on: true,
+          removeListener: true,
+          rpcEngine: false,
+        },
+        snap: {
+          request: true,
+          on: false,
+          removeListener: false,
+          rpcEngine: false,
+          requestType: 'function',
+        },
+      },
+    },
+    errors: undefined,
+  });
+});
+
+const testSubjects = ['ethereum', 'snap'];
+
+testSubjects.forEach((endowment) => {
+  test(`hardening protects ${endowment} endowment`, async (expect) => {
+    const CODE = `
+      module.exports.onRpcRequest = () => {
+        let result = 'ENDOWMENT_SECURED';
+        let errors = [];
+
+        try {
+          errors = (${testEndowmentHardening})(${endowment}, () => ${endowment});
+        } catch (error) {
+          return error.message;
+        }
+
+        try {
+          if (${endowment}.__flag) {
+            result = 'ENDOWMENT_NOT_SECURED';
+          }
+        } catch (error) {
+          return error.message;
+        }
+
+        try {
+          if (${endowment}.__proto__ && ${endowment}.__proto__.__flag) {
+            result = 'ENDOWMENT_NOT_SECURED';
+          }
+        } catch (error) {
+          return error.message;
+        }
+
+        try {
+          if (${endowment}.prototype && ${endowment}.prototype.__flag) {
+            result = 'ENDOWMENT_NOT_SECURED';
+          }
+        } catch (error) {
+          return error.message;
+        }
+
+        try {
+          const objectProto = Object.getPrototypeOf(${endowment});
+          if (objectProto.__flag) {
+            result = 'ENDOWMENT_NOT_SECURED';
+          }
+        } catch (error) {
+          return error.message;
+        }
```

Comment:
(nitpick) why have a separate try-catch for each if the catch is exactly the same _and_ it's stopping the execution anyway with a return?

Reply:
Good question actually. This was left after I've been doing some debugging and it was easier to do so, at the moment. I changed it now to be with single try/catch.